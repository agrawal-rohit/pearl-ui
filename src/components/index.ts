// Atoms
export { default as Box, BoxProps } from "./atoms/box/box";
export { default as Center } from "./atoms/center/center";
export { default as Collapse, CollapseProps } from "./atoms/collapse/collapse";
export { default as Divider, DividerProps } from "./atoms/divider/divider";
export { default as Fade, FadeProps } from "./atoms/fade/fade";
export { default as Grid, GridProps } from "./atoms/grid/grid";
export { default as Icon, IconProps } from "./atoms/icon/icon";
export {
  default as Pressable,
  PressableProps,
} from "./atoms/pressable/pressable";
export {
  default as ScaleFade,
  ScaleFadeProps,
} from "./atoms/scale-fade/scale-fade";
export {
  default as List,
  OrderedList,
  UnorderedList,
  ListProps,
} from "./atoms/list/list";
export { default as Screen, ScreenProps } from "./atoms/screen/screen";
export { default as Skeleton, SkeletonProps } from "./atoms/skeleton/skeleton";
export {
  default as SkeletonCircle,
  SkeletonCircleProps,
} from "./atoms/skeleton/skeleton-circle";
export {
  default as SkeletonText,
  SkeletonTextProps,
} from "./atoms/skeleton/skeleton-text";
export { default as Slide, SlideProps } from "./atoms/slide/slide";
export {
  default as SlideFade,
  SlideFadeProps,
} from "./atoms/slide-fade/slide-fade";
export { default as Spacer } from "./atoms/spacer/spacer";
export { default as Spinner, SpinnerProps } from "./atoms/spinner/spinner";
export {
  default as Stack,
  HStack,
  VStack,
  ZStack,
  HStackProps,
  VStackProps,
  StackProps,
} from "./atoms/stack/stack";
export { default as Text, TextProps } from "./atoms/text/text";

// Molecules
export { default as Avatar, AvatarProps } from "./molecules/avatar/avatar";
export {
  default as AvatarGroup,
  AvatarGroupProps,
} from "./molecules/avatar/avatar-group";
export { default as Badge, BadgeProps } from "./molecules/badge/badge";
export {
  default as withBadge,
  WithBadgeOptions,
} from "./molecules/badge/withBadge";
export { default as Button, ButtonProps } from "./molecules/button/button";
export {
  default as ButtonGroup,
  ButtonGroupProps,
} from "./molecules/button/button-group";
export {
  default as CheckBox,
  CheckBoxProps,
} from "./molecules/checkbox/checkbox";
export {
  default as CheckBoxGroup,
  CheckBoxGroupProps,
} from "./molecules/checkbox/checkbox-group";
export {
  default as IconButton,
  IconButtonProps,
} from "./molecules/icon-button/icon-button";
export { default as Image, ImageProps } from "./molecules/image/image";
export { default as Input, InputProps } from "./molecules/input/input";
export {
  default as PinInput,
  PinInputProps,
} from "./molecules/pin-input/pin-input";
export {
  default as Progress,
  ProgressProps,
} from "./molecules/progress/progress";
export { default as Radio, RadioProps } from "./molecules/radio/radio";
export {
  default as RadioGroup,
  RadioGroupProps,
} from "./molecules/radio/radio-group";
export { default as Switch, SwitchProps } from "./molecules/switch/switch";
export {
  default as Textarea,
  TextareaProps,
} from "./molecules/textarea/textarea";
export { default as Video, VideoProps } from "./molecules/video/video";
export {
  default as Accordion,
  AccordionProps,
  AccordionItemProps,
  AccordionButtonProps,
  AccordionPanelProps,
  AccordionItem,
  AccordionButton,
  AccordionPanel,
} from "./molecules/accordion/accordion";
export { default as Alert, AlertProps } from "./molecules/alert/alert";
