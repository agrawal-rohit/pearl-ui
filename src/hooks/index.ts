export { useColorModeValue } from "./useColorModeValue";
export { useComponentConfig } from "./useComponentConfig";
export { useStyledProps } from "./useStyledProps";
export { useTheme } from "./useTheme";
