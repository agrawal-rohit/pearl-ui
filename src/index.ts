export { default as Box } from "./components/Box/Box";
export { default as Button } from "./components/Button/Button";
export { default as Text } from "./components/Text/Text";
