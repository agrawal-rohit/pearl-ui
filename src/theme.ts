import { createTheme } from "@shopify/restyle";

const palette = {
  purpleLight: "#8C6FF7",
  purplePrimary: "#5A31F4",
  purpleDark: "#3F22AB",

  greenLight: "#56DCBA",
  greenPrimary: "#0ECD9D",
  greenDark: "#0A906E",

  black: "#0B0B0B",
  white: "#F0F2F3",
};

const theme = createTheme({
  colors: {
    primary: palette.purplePrimary,
    ...palette,
  },
  spacing: {
    s: 8,
    m: 16,
    l: 24,
    xl: 40,
  },
  breakpoints: {
    phone: 0,
    tablet: 768,
  },
});

export type Theme = typeof theme;
export default theme;
