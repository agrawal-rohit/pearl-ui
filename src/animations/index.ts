import { useAnimationState } from "../hooks/useAnimationState";

/**
 * Custom hook for creating a fade animation state.
 * This hook uses the useAnimationState hook to create an animation state that transitions from an opacity of 0 to an opacity of 1.
 * This creates a fade-in effect.
 *
 * @returns An animation state that creates a fade-in effect.
 */
export const useFade = () => {
  // Define the initial and final states for the fade animation.
  // The animation will start with an opacity of 0 (completely transparent) and end with an opacity of 1 (completely opaque).
  const fadeAnimationState = useAnimationState({
    from: {
      opacity: 0,
    },
    to: {
      opacity: 1,
    },
  });

  // Return the fade animation state.
  return fadeAnimationState;
};

/**
 * Custom hook for creating a scale and fade animation state.
 * This hook uses the useAnimationState hook to create an animation state that transitions from a scale of 0.5 and opacity of 0 to a scale of 1 and opacity of 1.
 * This creates a combined effect of scaling and fading in.
 *
 * @returns An animation state that creates a scale and fade-in effect.
 */
export const useScaleFade = () => {
  // Define the initial and final states for the scale and fade animation.
  // The animation will start with a scale of 0.5 (half the original size) and opacity of 0 (completely transparent)
  // and end with a scale of 1 (original size) and opacity of 1 (completely opaque).
  const scaleFadeAnimationState = useAnimationState({
    from: {
      scale: 0.9,
      opacity: 0,
    },
    to: {
      scale: 1,
      opacity: 1,
    },
  });

  // Return the scale and fade animation state.
  return scaleFadeAnimationState;
};

/**
 * Custom hook for creating a slide animation state.
 * This hook uses the useAnimationState hook to create an animation state that transitions from a translateX of -100 to a translateX of 0.
 * This creates a slide-in effect.
 *
 * @returns An animation state that creates a slide-in effect.
 */
export const useSlide = () => {
  // Define the initial and final states for the slide animation.
  // The animation will start with a translateX of -100 (completely off-screen to the left) and end with a translateX of 0 (original position).
  const slideAnimationState = useAnimationState({
    from: {
      translateY: -100,
    },
    to: {
      translateY: 0,
    },
  });

  // Return the slide animation state.
  return slideAnimationState;
};

/**
 * Custom hook for creating a slide and fade animation state.
 * This hook uses the useAnimationState hook to create an animation state that transitions from a translateX of -100 and opacity of 0 to a translateX of 0 and opacity of 1.
 * This creates a combined effect of sliding and fading in.
 *
 * @returns An animation state that creates a slide and fade-in effect.
 */
export const useSlideFade = () => {
  // Define the initial and final states for the slide and fade animation.
  // The animation will start with a translateX of -100 (completely off-screen to the left) and opacity of 0 (completely transparent)
  // and end with a translateX of 0 (original position) and opacity of 1 (completely opaque).
  const slideFadeAnimationState = useAnimationState({
    from: {
      opacity: 0,
      translateY: 10,
    },
    to: {
      opacity: 1,
      translateY: 0,
    },
  });

  // Return the slide and fade animation state.
  return slideFadeAnimationState;
};

/**
 * Custom hook for creating a collapse animation state.
 * This hook uses the useAnimationState hook to create an animation state that transitions from a scaleY of 1 to a scaleY of 0.
 * This creates a collapse effect.
 *
 * @returns An animation state that creates a collapse effect.
 */
export const useCollapse = () => {
  // Define the initial and final states for the collapse animation.
  // The animation will start with a scaleY of 1 (original size) and end with a scaleY of 0 (collapsed).
  const collapseAnimationState = useAnimationState({
    from: {
      opacity: 0,
      height: 0,
    },
    to: {
      opacity: 1,
      height: 100,
    },
  });

  // Return the collapse animation state.
  return collapseAnimationState;
};
