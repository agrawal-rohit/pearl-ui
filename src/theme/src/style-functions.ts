import { ColorValue, FlexStyle, TextStyle, ViewStyle } from "react-native";
import { getKeys, getNestedObject, isThemeKey } from "../utils/type-helpers";
import {
  borderColorProperties,
  borderProperties,
  borderRadiusProperties,
  layoutProperties,
  layoutPropertiesShorthand,
  positionProperties,
  shadowProperties,
  spacingProperties,
  spacingPropertiesShorthand,
  textShadowProperties,
  typographyProperties,
} from "./style-properties";
import {
  FinalPearlTheme,
  ColorModeColor,
  ResponsiveValue,
  RNStyleProperty,
  StyleFunction,
  StyleFunctionContainer,
  PaletteColors,
} from "./types";
import { useTheme } from "../../hooks/useTheme";
import {
  getValueForScreenSize,
  isResponsiveObjectValue,
} from "./responsive-helpers";

interface CreateStyleFunctionProps {
  property: string;
  transform?: any;
  styleProperty?: RNStyleProperty | "skip" | string;
  themeKey?: keyof FinalPearlTheme | undefined;
}

/**
 * Creates a style function container that can be used to create a style function.
 * @param property - The name of the property to create a style function for.
 * @param transform - An optional transformation function to apply to the value of the property.
 * @param styleProperty - An optional name of the style property to use instead of the property name.
 * @param themeKey - An optional name of the theme key to use for the property value.
 * @returns A style function container that can be used to create a style function.
 */
export const createStyleFunction = ({
  property,
  transform,
  styleProperty,
  themeKey,
}: CreateStyleFunctionProps): StyleFunctionContainer => {
  const styleProp = styleProperty ?? property.toString();

  /**
   * A style function that returns a style object with the given property and value.
   * @param props - The props object to get the property value from.
   * @param theme - The theme object to get the theme value from.
   * @param dimensions - The dimensions object to get the screen size from.
   * @returns A style object with the given property and value.
   */
  const func: StyleFunction = (props: any, { theme, dimensions }: any): any => {
    // Initial value is the raw prop value
    let value = isResponsiveObjectValue(props[property], theme)
      ? getValueForScreenSize({
          responsiveValue: props[property],
          breakpoints: theme.breakpoints,
          dimensions,
        })
      : props[property];

    // Transform the value if transformation function exists
    if (transform) {
      value = transform(value);
    }

    // Check if this value refers to a key in the theme config
    if (isThemeKey(theme, themeKey)) {
      if (value) {
        let themeTokenValue = theme[themeKey][value];

        // For color palettes with multiple shades
        if (typeof value === "string" && value.match(/^[a-z]+\.\d+$/)) {
          themeTokenValue = getNestedObject(theme, [
            themeKey,
            value.split(".")[0],
            value.split(".")[1],
          ]);
        }

        // Apply the value
        value = themeTokenValue ?? value;
      }
    }

    if (value === undefined) return {};

    if (styleProperty !== "skip") {
      return {
        [styleProp]: value,
      };
    } else {
      return value;
    }
  };

  return {
    property,
    themeKey,
    func,
  };
};

/**
 * Transforms a color value based on the current color mode.
 * @param value - The color value to transform.
 * @returns The transformed color value.
 */
export const transformColorValue = (value: ColorModeColor | ColorValue) => {
  const { colorMode } = useTheme();

  // Color Mode color provided
  if (typeof value === "object") {
    if (colorMode === "light") return value.light;

    return value.dark;
  }

  return value;
};

export const backgroundColorStyleFunction = [
  createStyleFunction({
    property: "backgroundColor",
    themeKey: "palette",
    transform: transformColorValue,
  }),
  createStyleFunction({
    property: "bgColor",
    styleProperty: "backgroundColor",
    themeKey: "palette",
    transform: transformColorValue,
  }),
];

export const colorStyleFunction = createStyleFunction({
  property: "color",
  themeKey: "palette",
  transform: transformColorValue,
});

export const opacityStyleFunction = createStyleFunction({
  property: "opacity",
});

export const visibleStyleFunction = createStyleFunction({
  property: "visible",
  styleProperty: "opacity",
  transform: (value: any) =>
    value !== undefined ? (value === false ? 0 : 1) : undefined,
});

export const typographyStyleFunction = [
  ...getKeys(typographyProperties).map((property) => {
    return createStyleFunction({
      property,
    });
  }),
  createStyleFunction({
    property: "letterSpacing",
    themeKey: "letterSpacings",
  }),
  createStyleFunction({
    property: "fontFamily",
    themeKey: "fonts",
  }),
  createStyleFunction({
    property: "fontWeight",
    themeKey: "fontWeights",
  }),
  createStyleFunction({
    property: "fontSize",
    themeKey: "fontSizes",
  }),
  createStyleFunction({
    property: "lineHeight",
    themeKey: "lineHeights",
  }),
];

export const layoutStyleFunction = [
  ...getKeys(layoutProperties).map((property) => {
    return createStyleFunction({
      property,
    });
  }),
  ...getKeys(layoutPropertiesShorthand).map((property) => {
    const styleProperty = layoutPropertiesShorthand[property];

    return createStyleFunction({
      property,
      styleProperty,
    });
  }),
];

export const spacingStyleFunction = [
  ...getKeys(spacingProperties).map((property) => {
    return createStyleFunction({
      property,
      themeKey: "spacing",
    });
  }),
  ...getKeys(spacingPropertiesShorthand).map((property) => {
    const styleProperty = spacingPropertiesShorthand[property];
    return createStyleFunction({
      property,
      styleProperty,
      themeKey: "spacing",
    });
  }),
];

export const positionStyleFunction = [
  ...getKeys(positionProperties).map((property) => {
    return createStyleFunction({
      property,
    });
  }),
  createStyleFunction({
    property: "zIndex",
    themeKey: "zIndices",
  }),
];

export const borderStyleFunction = [
  ...getKeys(borderProperties).map((property) => {
    return createStyleFunction({
      property,
    });
  }),
  ...getKeys(borderColorProperties).map((property) => {
    return createStyleFunction({
      property,
      themeKey: "palette",
      transform: transformColorValue,
    });
  }),
  ...getKeys(borderRadiusProperties).map((property) => {
    return createStyleFunction({
      property,
      themeKey: "borderRadii",
    });
  }),
];

export const shadowStyleFunction = [
  ...getKeys(shadowProperties).map((property) => {
    return createStyleFunction({
      property,
    });
  }),
  createStyleFunction({
    property: "boxShadow",
    themeKey: "elevation",
    styleProperty: "skip",
  }),
  createStyleFunction({
    property: "shadowColor",
    themeKey: "palette",
    transform: transformColorValue,
  }),
];

export const textShadowStyleFunction = [
  ...getKeys(textShadowProperties).map((property) => {
    return createStyleFunction({
      property,
    });
  }),
  createStyleFunction({
    property: "textShadowColor",
    themeKey: "palette",
    transform: transformColorValue,
  }),
];

export const allStyleFunctions = [
  colorStyleFunction,
  opacityStyleFunction,
  visibleStyleFunction,
  ...backgroundColorStyleFunction,
  ...spacingStyleFunction,
  ...typographyStyleFunction,
  ...layoutStyleFunction,
  ...positionStyleFunction,
  ...borderStyleFunction,
  ...shadowStyleFunction,
  ...textShadowStyleFunction,
];

// Component Specific Style function
export const boxStyleFunctions = [
  opacityStyleFunction,
  visibleStyleFunction,
  ...backgroundColorStyleFunction,
  ...layoutStyleFunction,
  ...spacingStyleFunction,
  ...borderStyleFunction,
  ...shadowStyleFunction,
  ...positionStyleFunction,
] as StyleFunctionContainer[];

export const textStyleFunctions = [
  colorStyleFunction,
  backgroundColorStyleFunction,
  opacityStyleFunction,
  visibleStyleFunction,
  layoutStyleFunction,
  typographyStyleFunction,
  spacingStyleFunction,
  textShadowStyleFunction,
  positionStyleFunction,
] as StyleFunctionContainer[];

// PropTypes
export interface ColorProps {
  color?: ResponsiveValue<PaletteColors>;
}
export interface OpacityProps {
  opacity?: ResponsiveValue<number>;
}

export interface VisibleProps {
  visible?: ResponsiveValue<boolean>;
}

export interface BackgroundColorProps {
  backgroundColor?: ResponsiveValue<PaletteColors>;
  bgColor?: ResponsiveValue<PaletteColors>;
}

type SpacingPropsBase = {
  [Key in keyof typeof spacingProperties]?: ResponsiveValue<
    keyof FinalPearlTheme["spacing"]
  >;
};

type SpacingShorthandProps = {
  [Key in keyof typeof spacingPropertiesShorthand]?: ResponsiveValue<
    keyof FinalPearlTheme["spacing"]
  >;
};

export type SpacingProps = SpacingPropsBase & SpacingShorthandProps;

export type TypographyProps = {
  [Key in keyof typeof typographyProperties]?: TextStyle[Key];
} & {
  letterSpacing?: ResponsiveValue<keyof FinalPearlTheme["letterSpacings"]>;
  fontSize?: ResponsiveValue<keyof FinalPearlTheme["fontSizes"]>;
  lineHeight?: ResponsiveValue<keyof FinalPearlTheme["lineHeights"]>;
  fontWeight?: ResponsiveValue<keyof FinalPearlTheme["fontWeights"]>;
  fontFamily?: ResponsiveValue<keyof FinalPearlTheme["fonts"]>;
};

type LayoutPropsBase = {
  [Key in keyof typeof layoutProperties]?: ResponsiveValue<FlexStyle[Key]>;
};

type LayoutShorthandProps = {
  [Key in keyof typeof layoutPropertiesShorthand]?: ResponsiveValue<
    FlexStyle[(typeof layoutPropertiesShorthand)[Key]]
  >;
};

export type LayoutProps = LayoutPropsBase & LayoutShorthandProps;

export type PositionProps = {
  [Key in keyof typeof positionProperties]?: ResponsiveValue<FlexStyle[Key]>;
} & {
  zIndex?: ResponsiveValue<keyof FinalPearlTheme["zIndices"]>;
};

export type BorderProps = {
  [Key in keyof typeof borderProperties]?: ResponsiveValue<ViewStyle[Key]>;
} & {
  [Key in keyof typeof borderColorProperties]?: ResponsiveValue<PaletteColors>;
} & {
  [Key in keyof typeof borderRadiusProperties]?: ResponsiveValue<
    keyof FinalPearlTheme["borderRadii"]
  >;
};

export type ShadowProps = {
  [Key in keyof typeof shadowProperties]?: ResponsiveValue<ViewStyle[Key]>;
} & {
  boxShadow?: ResponsiveValue<keyof FinalPearlTheme["elevation"]>;
  shadowColor?: ResponsiveValue<PaletteColors>;
};

export type TextShadowProps = {
  [Key in keyof typeof textShadowProperties]?: ResponsiveValue<TextStyle[Key]>;
} & {
  textShadowColor?: ResponsiveValue<PaletteColors>;
};

export type AllProps = BackgroundColorProps &
  ColorProps &
  OpacityProps &
  SpacingProps &
  TypographyProps &
  LayoutProps &
  PositionProps &
  BorderProps &
  ShadowProps &
  TextShadowProps;

// Component Style Prop
export type BoxStyleProps = BackgroundColorProps &
  OpacityProps &
  VisibleProps &
  LayoutProps &
  SpacingProps &
  BorderProps &
  ShadowProps &
  PositionProps;

export type TextStyleProps = ColorProps &
  BackgroundColorProps &
  OpacityProps &
  VisibleProps &
  LayoutProps &
  TypographyProps &
  SpacingProps &
  TextShadowProps;
