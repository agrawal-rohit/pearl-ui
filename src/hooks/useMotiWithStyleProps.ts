import { useMemo } from "react";
import { getKeys } from "../theme/utils/type-helpers";
import { StyleFunctionContainer } from "../theme/src/types";
import { useTheme } from "./useTheme";
import { composeCleanStyleProps, composeStyleProps } from "./utils/utils";
import {
  layoutPropertiesShorthand,
  spacingPropertiesShorthand,
} from "../theme/src/style-properties";
import _ from "lodash";
import { removeUndefined } from "../theme/utils/utils";
import { useWindowDimensions } from "react-native";

const shorthandPropMapper = {
  ...layoutPropertiesShorthand,
  ...spacingPropertiesShorthand,
};

/**
 * Converts Moti Props using style props as well
 * @param props - The props to convert
 * @param styleFunctions - The style functions to use
 * @returns The converted props
 */
export const useMotiWithStyleProps = (
  props: Record<string, any>,
  styleFunctions: StyleFunctionContainer[]
) => {
  const { theme, colorMode } = useTheme();
  const dimensions = useWindowDimensions();

  // Compose style properties
  const buildStyleProperties = useMemo(
    () => composeStyleProps(styleFunctions),
    [styleFunctions]
  );

  const nonAnimateableStyleProps = [
    "flex",
    "flexShrink",
    "flexGrow",
    "flexBasis",
    "flexDirection",
    "flexWrap",
    "shadowOffset",
    "textShadowOffset",
    "fontFamily",
    "fontWeight",
    "fontSize",
    "lineHeight",
    "fontStyle",
    "textAlign",
    "letterSpacing",
    "textDecorationLine",
    "textDecorationStyle",
    "textTransform",
  ];
  const animateableStyles = _.omit(props.style, nonAnimateableStyleProps);

  // Convert 'from', 'animate', 'transition', 'exit', 'exitTransition' props
  ["from", "animate", "transition", "exit", "exitTransition"].forEach(
    (prop) => {
      if (prop === "animate" && props[prop] === undefined) props[prop] = {};

      if (props[prop]) {
        props[prop] = composeCleanStyleProps(
          props[prop],
          buildStyleProperties,
          {
            theme,
            colorMode,
            dimensions,
          }
        );

        if (prop === "animate") {
          // Merge animateableStyles and animate
          props.animate = {
            ...animateableStyles,
            ...removeUndefined(props.animate),
          };
        }

        if (["transition", "exitTransition"].includes(prop)) {
          // Filter object values from 'transition' or 'exitTransition'
          const keysWithObjectValues = getKeys(props[prop]).filter(
            (key) => typeof props[prop][key] === "object"
          );
          const propsWithObjectValues = _.pick(
            props[prop],
            keysWithObjectValues
          );
          const nullObject = keysWithObjectValues.reduce((obj, key) => {
            return {
              ...obj,
              [key]: null,
            };
          }, {});
          props[prop] = { ...props[prop], ...removeUndefined(nullObject) };

          // Add the filtered values back into 'transition' or 'exitTransition'
          props[prop] = keysWithObjectValues.reduce((obj, key: any) => {
            let finalKey = key;
            if (getKeys(shorthandPropMapper).includes(key)) {
              finalKey = (shorthandPropMapper as any)[key];
            }

            return {
              ...obj,
              [finalKey]: propsWithObjectValues[key],
            };
          }, props[prop]);
        }
      }
    }
  );

  // props.style = _.pick(props.style, nonAnimateableStyleProps);
  return props;
};
