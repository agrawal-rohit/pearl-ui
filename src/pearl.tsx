import {
  AtomComponent,
  ComponentTypes,
  FinalPearlTheme,
  MoleculeComponent,
  PearlComponent,
  StyleFunctionContainer,
} from "./theme/src/types";
import React from "react";
import { useAtomicComponentConfig } from "./hooks/useAtomicComponentConfig";
import { boxStyleFunctions, BoxStyleProps } from "./theme/src/style-functions";
import { useStyleProps } from "./hooks/useStyleProps";
import { useMolecularComponentConfig } from "./hooks/useMolecularComponentConfig";
import { MotiView, motify } from "moti";
import { useMotiWithStyleProps } from "./hooks/useMotiWithStyleProps";
import { useTheme } from "./hooks/useTheme";
import { Platform } from "react-native";

/**
 * Configuration object that adds metadata to a component that's required by Pearl UI
 */
interface pearlConfig {
  /**
   * The name of the component in the FinalPearlTheme
   */
  componentName: keyof FinalPearlTheme["components"];
  /**
   * The type of the component
   */
  type: ComponentTypes;
  /**
   * Whether the component is animatable or not
   */
  animatable?: boolean;
}

/**
 * Hook to convert a third-party component to a Pearl component that can be configured using style props, an atomic component configuration or a molecular component configuration.
 * @param Component The base component on which the pearlification has to be applied
 * @param config A configuration object that adds metadata to a component that's required by Pearl UI
 * @param styleFunctions The set of style functions that would define the style props that should be supported by the pearlified component
 * @param moleculeConfigOptions The options for configuring the molecular component
 * @returns
 */
export function pearl<
  ComponentProps,
  ComponentType extends ComponentTypes = "basic",
  ComponentAtoms extends Record<string, any> = Record<string, any>,
  StyleProps = BoxStyleProps,
  Animateable extends boolean = true,
>(
  Component: any,
  config: pearlConfig = {
    componentName: "None",
    type: "basic",
    animatable: true,
  },
  styleFunctions: StyleFunctionContainer[] = boxStyleFunctions,
  moleculeConfigOptions: {
    partForOverridenStyleProps?: string | undefined;
    partForOverridenNativeProps?: string | undefined;
    partForOverridenAnimationProps?: string | undefined;
  } = {}
) {
  let FinalComponent: any | undefined;
  const isAnimatable = !!config.animatable && config.type !== "molecule";

  if (isAnimatable) {
    const ConvertedClassComponent = class extends React.Component<any, any> {
      static displayName = `Pearl${
        Component.name ?? Component.displayName ?? `NoName`
      }`;

      render() {
        const { children, ...props } = this.props;
        return (
          <Component ref={(props as any).forwardedRef} {...props}>
            {children}
          </Component>
        );
      }
    };

    if (Platform.OS === "web" && config.componentName === "Box")
      FinalComponent = MotiView;
    else FinalComponent = motify(ConvertedClassComponent)();
  } else {
    FinalComponent = Component;
  }

  return React.memo(
    React.forwardRef(
      (
        {
          children,
          ...rest
        }: PearlComponent<ComponentProps, StyleProps, Animateable> &
          AtomComponent<(typeof config)["componentName"], ComponentType> &
          MoleculeComponent<(typeof config)["componentName"], ComponentType> & {
            atoms?: Partial<ComponentAtoms>;
            children?: string | JSX.Element | JSX.Element[] | React.ReactNode;
          },
        ref: any
      ) => {
        const { colorMode } = useTheme();

        let convertedProps;
        if (config.type === "atom") {
          convertedProps = useAtomicComponentConfig(
            config["componentName"],
            rest,
            {
              size: (rest as any).size,
              variant: (rest as any).variant,
            },
            rest.colorScheme,
            styleFunctions
          );
          if (config.animatable)
            convertedProps = useMotiWithStyleProps(
              convertedProps,
              styleFunctions
            );
        } else if (config.type === "molecule") {
          convertedProps = useMolecularComponentConfig(
            config["componentName"],
            rest,
            {
              size: (rest as any).size,
              variant: (rest as any).variant,
            },
            rest.colorScheme,
            styleFunctions,
            moleculeConfigOptions.partForOverridenStyleProps,
            moleculeConfigOptions.partForOverridenNativeProps,
            moleculeConfigOptions.partForOverridenAnimationProps
          );
        } else {
          convertedProps = useStyleProps(rest, styleFunctions);
          if (config.animatable) {
            convertedProps = useMotiWithStyleProps(
              convertedProps,
              styleFunctions
            );
          }
        }

        return (
          <FinalComponent
            {...convertedProps}
            key={
              Platform.OS === "web" &&
              ["Text", "Icon", "Box"].includes(config.componentName as string)
                ? (convertedProps as any).key!
                  ? `${(convertedProps as any).key!}-${colorMode}`
                  : `${colorMode}`
                : undefined
            }
            ref={ref}
          >
            {children}
          </FinalComponent>
        );
      }
    )
  );
}
