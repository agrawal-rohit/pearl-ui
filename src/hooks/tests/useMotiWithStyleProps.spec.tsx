import React from "react";
import { render } from "@testing-library/react-native";
import { ThemeProvider } from "../../theme/src/theme-context";
import { allStyleFunctions } from "../../theme/src/style-functions";
import { useMotiWithStyleProps } from "../useMotiWithStyleProps";
import { MotiView } from "moti";

jest.useFakeTimers();

const TestComponent: React.FC<any> = (props: any) => {
  const rest = useMotiWithStyleProps(props.object, allStyleFunctions);

  return <MotiView {...rest}></MotiView>;
};

describe("useMotiWithStyleProps", () => {
  it("transforms Moti 'from' props correctly", () => {
    const comp = render(
      <ThemeProvider>
        <TestComponent object={{ from: { backgroundColor: "pink", p: "s" } }} />
      </ThemeProvider>
    );

    const tree = comp.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("transforms Moti 'animate' props correctly", () => {
    const comp = render(
      <ThemeProvider>
        <TestComponent
          object={{ animate: { backgroundColor: "pink", p: "s" } }}
        />
      </ThemeProvider>
    );

    const tree = comp.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("transforms Moti 'exit' props correctly", () => {
    const comp = render(
      <ThemeProvider>
        <TestComponent object={{ exit: { backgroundColor: "pink", p: "s" } }} />
      </ThemeProvider>
    );

    const tree = comp.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("transforms Moti 'transition' props correctly", () => {
    const comp = render(
      <ThemeProvider>
        <TestComponent object={{ transition: { type: "timing" } }} />
      </ThemeProvider>
    );

    const tree = comp.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("transforms Moti 'exitTransition' props correctly", () => {
    const comp = render(
      <ThemeProvider>
        <TestComponent object={{ exitTransition: { type: "timing" } }} />
      </ThemeProvider>
    );

    const tree = comp.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("transforms Moti 'state' props correctly", () => {
    const comp = render(
      <ThemeProvider>
        <TestComponent
          object={{
            state: {
              from: { backgroundColor: "pink", p: "s" },
              to: { backgroundColor: "silver", p: "xs" },
            },
          }}
        />
      </ThemeProvider>
    );

    const tree = comp.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
