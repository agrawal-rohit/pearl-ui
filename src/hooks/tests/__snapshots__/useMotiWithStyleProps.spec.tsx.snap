// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useMotiWithStyleProps transforms Moti 'animate' props correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    animate={
      {
        "backgroundColor": "#ffc0cb",
        "padding": "s",
      }
    }
    animatedStyle={
      {
        "value": {
          "backgroundColor": "#ffc0cb",
          "padding": "s",
        },
      }
    }
    collapsable={false}
    style={
      {
        "backgroundColor": "#ffc0cb",
        "padding": "s",
      }
    }
  />
</View>
`;

exports[`useMotiWithStyleProps transforms Moti 'exit' props correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    animate={{}}
    animatedStyle={
      {
        "value": {},
      }
    }
    collapsable={false}
    exit={
      {
        "backgroundColor": "#ffc0cb",
        "padding": "s",
      }
    }
    style={{}}
  />
</View>
`;

exports[`useMotiWithStyleProps transforms Moti 'exitTransition' props correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    animate={{}}
    animatedStyle={
      {
        "value": {},
      }
    }
    collapsable={false}
    exitTransition={
      {
        "type": "timing",
      }
    }
    style={{}}
  />
</View>
`;

exports[`useMotiWithStyleProps transforms Moti 'from' props correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    animate={{}}
    animatedStyle={
      {
        "value": {
          "backgroundColor": "#ffc0cb",
          "padding": "s",
        },
      }
    }
    collapsable={false}
    from={
      {
        "backgroundColor": "#ffc0cb",
        "padding": "s",
      }
    }
    style={
      {
        "backgroundColor": "#ffc0cb",
        "padding": "s",
      }
    }
  />
</View>
`;

exports[`useMotiWithStyleProps transforms Moti 'state' props correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    animate={{}}
    animatedStyle={
      {
        "value": {},
      }
    }
    collapsable={false}
    state={
      {
        "from": {
          "backgroundColor": "pink",
          "p": "s",
        },
        "to": {
          "backgroundColor": "silver",
          "p": "xs",
        },
      }
    }
    style={{}}
  />
</View>
`;

exports[`useMotiWithStyleProps transforms Moti 'transition' props correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    animate={{}}
    animatedStyle={
      {
        "value": {},
      }
    }
    collapsable={false}
    style={{}}
    transition={
      {
        "type": "timing",
      }
    }
  />
</View>
`;
